### PR TITLE
[stable/jenkins] Deprecating some values in favour of JCasC

### PR DESCRIPTION
**Pull request by holmesb**
Addresses issue #11276.  Replaces PR 11691 which was a mess of historical commits.

### CODE REVIEW

**Comment by maorfr** [comment]
/assign
/ok-to-test

I will take a close look at all the changes a bit later and let the tests run for now.

Do you think that this should be a major version bump?

**Comment by holmesb** [comment]
> major version bump?

Not sure.  In terms of user experience, nothing should change beyond the welcome message.  Only reason I major-bumped is we are enabling CasC by default for the first time.  Hopefully this will be the point at which little/no further development will occur in the unholy-mess that is config.yaml and instead users can add their own configuration in values.yaml under Master.JCasC.ConfigScripts.  In terms of this chart's development, I'd say it's significant enough.

File: stable/jenkins/values.yaml [hunk 1]
```diff
@@ -173,13 +164,62 @@ Master:
...
+        security:
+            remotingCLI:
+                  enabled: false
+      kubernetes-slave: |
+        jenkins:
+          clouds:
+            - kubernetes:
+                name: "jenkins-slave"
+                serverUrl: https://kubernetes.default
+                skipTlsVerify: true
```
Comment (torstenwalter) [inline stable/jenkins/values.yaml]: This should be `skipTlsVerify: false` as it was before: https://github.com/helm/charts/blob/master/stable/jenkins/templates/config.yaml#L131

File: stable/jenkins/values.yaml [hunk 2]
```diff
@@ -173,13 +164,62 @@ Master:
...
+                maxRequestsPerHostStr: 64
+                retentionTimeout: 5
+                templates:
+                  - name: "default"
+                    label: "jenkins-slave"
+                    idleMinutes: "1"
+                    activeDeadlineSeconds: "120"
+                    slaveConnectTimeout: "1000"
+                    volumes:
+                      - secretVolume:
```
Comment (torstenwalter) [inline stable/jenkins/values.yaml]: Please remove the `secretVolume`. It's probaly a copy and pase error. Agent won't start if the secret `mysecret` does not exist.

File: stable/jenkins/values.yaml [hunk 3]
```diff
@@ -173,13 +164,62 @@ Master:
...
+                          secretName: mysecret
+                          mountPath: "/var/myapp/mysecret"
+                    containers:
+                      - name: "jnlp"
+                        image: "jenkinsci/jnlp-slave:3.27-1"
+                        alwaysPullImage: true
```
Comment (torstenwalter) [inline stable/jenkins/values.yaml]: should be `alwaysPullImage: false` (https://github.com/helm/charts/blob/master/stable/jenkins/values.yaml#L287)

It does not make sense to always pull the image as the tag won't change.

File: stable/jenkins/values.yaml (hunk 1 above)
Comment (holmesb) [reply stable/jenkins/values.yaml]: Addressed in [commit](https://github.com/helm/charts/pull/11706/commits/e0faa81d38915f54a1c39c760563ba7ce7e04503)

File: stable/jenkins/values.yaml (hunk 2 above)
Comment (holmesb) [reply stable/jenkins/values.yaml]: Addressed in [commit](https://github.com/helm/charts/pull/11706/commits/e0faa81d38915f54a1c39c760563ba7ce7e04503)

File: stable/jenkins/values.yaml (hunk 3 above)
Comment (holmesb) [reply stable/jenkins/values.yaml]: Addressed in [commit](https://github.com/helm/charts/pull/11706/commits/e0faa81d38915f54a1c39c760563ba7ce7e04503)

**Comment by maorfr** [comment]
@holmesb,

I have given this PR a lot of thought.
I am very glad about improving this chart!

There is one thing that is bugging me though, and I would appreciate your thoughts.

I always felt like `values.yaml` should be as simple and as lean as possible, and not just another container for configuration. As a user, I want a simple experience. For example, if I want the exact stable chart, but with a different slave image, this PR will make it harder then simply using `.Agent.Image`.

I would suggest to keep `values.yaml` more similar to its current state (I'm always +1 for simplifying it), and add the configuration complexity into the templates.

What do you think?

**Comment by holmesb** [comment]
Hi @maorfr sorry for the late reply, I'm away snowboarding this week.  I don't think templating CasC is feasible.  Keeping this chart in sync with the latest CasC demos & documentation would be too much work and error prone.  Would always be a losing battle.  Would only work if this chart was made part of the [CasC project](https://github.com/jenkinsci/configuration-as-code-plugin) so maintainers there can keep it up to date.  Perhaps Jenkins in this repo could just be a link to the [helm](https://github.com/jenkinsci/configuration-as-code-plugin/tree/master/demos/kubernetes-helm) section there.

Simplicity IMO, is giving users a single method of configuring.  Without deprecating these values, we'll have two: 1) config.xml (and others) munging 2) CasC.  This is fine from my team's PoV, since we'll just continue using the latter.,  But it means the [simplification ](https://github.com/helm/charts/pull/9057#issuecomment-461722527) some have been asking for (@JeanMertz) won't be delivered.  Perhaps it's easier for people to choose between a second, pared-down "Jenkins-with-CasC" chart (perhaps with templated CasC that is linked to the CasC repo) and this one than deprecate 1) in this chart.  Not sure if having two Jenkins charts in this repo is orthodox\allowed.

**Comment by viglesiasce** [comment]
I agree with @maorfr that we should leave in the old values for now and maybe start thinking about making a JCasC only version of this chart (v2?) that attempts to simplify things back down.

**Comment by tjanson** [comment]
@holmesb 

> I don't think templating CasC is feasible. Keeping this chart in sync with the latest CasC demos & documentation would be too much work and error prone. Would always be a losing battle.

Just out of curiosity: How is templating the CasC YAML config more error prone than templating the “native” XML config? I don’t see the difference (both are at the mercy of backwards-incompatible API changes, if there ever are any). Am I missing something?

**Comment by holmesb** [comment]
Hi @tjanson, I wanted to move the responsibility for declaring configuration from developers of this chart to the users.  I believe CasC syntax is simple enough that users can copy & paste the demos from the [CasC github](https://github.com/jenkinsci/configuration-as-code-plugin) into their values.yaml and change to suit.  This chart should aim to give sensible default settings and empower users to quickly get up to speed with CasC.  Arguably perhaps also a small amount of templated CasC for a few of the most common use-cases.  But micro-defining every aspect of Jenkins is better done by the CasC project IMO, otherwise this chart will become overly-complex and difficult to maintain (like confg.yaml is now).

Also, at the time, CasC code and configuration syntax was being fast-developed and was more dynamic than the "native" config, so templating risked breakages.  I'm not close enough to the CasC project to know whether this aspect of my rationale is still valid.

**Pull request closed without merging**